clash-lib/src/Clash/Normalize/Util.hs

```diff
@@ -480,25 +477,22 @@ tvFirst _ = Nothing
 tvSubstWithTyEq
   :: Type
   -> Type
-tvSubstWithTyEq ty0 = go [] args0
+tvSubstWithTyEq ty0 = go [] False ty0
  where
-  (args0,tyRes) = splitFunForallTy ty0
-
-  go :: [(TyVar,Type)] -> [Either TyVar Type] -> Type
-  go eqs (Right arg : args)
+  go :: [Either TyVar Type] -> Bool -> Type -> Type
+  go argsOut changed (ForAllTy tv ty)
+    = go (Left tv:argsOut) changed ty
+  go argsOut changed (tyView -> FunTy arg tyRes)
     | Just (tc,tcArgs) <- splitTyConAppM arg
     , nameUniq tc == getKey eqTyConKey
-    , Just eq <- tvFirst tcArgs
-    = go (eq:eqs) args
-    | otherwise = go eqs args
-  go eqs (Left _tv : args)
-    = go eqs args -- drop (ForAll) tv
-  go []  [] = ty0 -- no eq constraints, returning original type
-  go eqs [] = substTy subst ty2
-   where
-     subst = extendTvSubstList (mkSubst emptyInScopeSet) eqs
-     args2 = args0 List.\\ (map (Left . fst) eqs)
-     ty2 = mkPolyFunTy tyRes args2
+    , Just (tv,ty) <- tvFirst tcArgs
+    = let
+        subst = extendTvSubst (mkSubst $ mkInScopeSet $ mkVarSet $ lefts argsOut2) tv ty
+        argsOut2 = Right arg : (argsOut List.\\ [Left tv])
```

Comment:
Perhaps reverse these two statements? I know the order doesn't matter, but I still read let-expressions from top to bottom.